fedimint-core/src/encoding/mod.rs

```diff
@@ -795,16 +883,25 @@ where
 
 impl<K> Decodable for BTreeSet<K>
 where
-    K: Decodable + Ord,
+    K: Decodable + Ord + Clone,
 {
-    fn consensus_decode<D: std::io::Read>(
+    fn consensus_decode_from_finite_reader<D: std::io::Read>(
         d: &mut D,
         modules: &ModuleDecoderRegistry,
     ) -> Result<Self, DecodeError> {
         let mut res = BTreeSet::new();
-        let len = u64::consensus_decode(d, modules)?;
+        let len = u64::consensus_decode_from_finite_reader(d, modules)?;
+        let mut prev_key = None;
         for _ in 0..len {
-            let k = K::consensus_decode(d, modules)?;
+            let k = K::consensus_decode_from_finite_reader(d, modules)?;
+            if prev_key
+                .as_ref()
+                .map(|prev_key| &k <= prev_key)
+                .unwrap_or_default()
+            {
+                return Err(DecodeError::from_str("Non-canonical encoding"));
+            }
+            prev_key = Some(k.clone());
```

Comment:
I'm wondering if we should imple this as `BTreeMap<K, ()>::consensus_decode_from_finite_reader(…)`.